internal/errors/errors.go

```diff
@@ -135,3 +135,21 @@ func IsErrMissingRequiredSecret(err error) bool {
 	_, ok := err.(ErrMissingRequiredSecret)
 	return ok
 }
+
+type ErrInvalidConfiguration struct {
+	message string
+}
+
+func NewErrInvalidConfiguration(message string) ErrInvalidConfiguration {
+	return ErrInvalidConfiguration{message: message}
+}
```

Comment:
All the callers have an error, so I believe this should take `cause error` as an argument and do usual `errors.Unwrap` go stuff.

Reply:
After reworking the individual module apply methods to raise configuration errors, this is not necessarily the case anymore.

Reply:
I think my comment got misunderstood, since this is now `return ErrInvalidConfiguration{message: cause.Error()}`.

What I meant by "usual `errors.Unwrap` go stuff" was the following boilerplate:

```go
type ErrInvalidConfiguration struct {
	inner error
}

func NewErrInvalidConfiguration(cause error) error { return ErrInvalidConfiguration{cause} }

func (e ErrInvalidConfiguration) Unwrap() error {
    return e.inner
}
```

Stringifying it loses the inner type. If we want to intentionally lose the inner type, the right way to do that is `errors.Opaque(err)`, which signals intent more specifically, but in this case, I don't think we actually want to lose the inner type.

In addition, the return type should be `error`, not `ErrInvalidConfiguration`, since otherwise type inference will make it annoying if you name it err and try to reuse it, i.e.:

```go
err := NewErrInvalidConfiguration()

// later in the same scope

ret, err := someOtherCall() // compiler error, cannot use err (value of type error) as ErrInvalidConfiguration value in assignment
```

Go's error handling basically requires for you to do runtime type-assertions only (errors.As etc), and at compile-time to squash all types down to just `error`. It's ugly, but with the type system and warts go has, the other options are worse sadly.

Reply:
Ditto'ing [comment here](https://github.com/ngrok/kubernetes-ingress-controller/pull/243#discussion_r1223044376). Went ahead and changed it as specified.